File: design-hard/297.二叉树的序列化与反序列化.cpp
/*
 * @lc app=leetcode.cn id=297 lang=cpp
 *
 * [297] 二叉树的序列化与反序列化
 */
#include <bits/stdc++.h>

#include "../TreeNode.cpp"
using namespace std;
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:
    // Encodes a tree to a single string.
    string serialize(TreeNode *root) {
        if (root == NULL)
            return "#";
        return to_string(root->val) + "," + serialize(root->left) + "," + serialize(root->right);
    }
    TreeNode *generate(const string data, int &pos) {
        if (data[pos] == '#') {
            pos += 2;
            return NULL;
        }
        int next = pos;
        while (data[next] != ',')
            next++;
        TreeNode *root = new TreeNode(stoi(data.substr(pos, next - pos)));
        pos = next + 1;
        root->left = generate(data, pos);
        root->right = generate(data, pos);
        return root;
    }
    // Decodes your encoded data to tree.
    TreeNode *deserialize(string data) {
        int pos = 0;
        return generate(data, pos);
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));
// @lc code=end

File: tree-hard/297.二叉树的序列化与反序列化.py
#
# @lc app=leetcode.cn id=297 lang=python3
#
# [297] 二叉树的序列化与反序列化
#
import collections


class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return '[]'
        ret = list()
        queue = collections.deque()
        queue.append(root)
        while queue:
            node = queue.popleft()
            if node:
                ret.append(str(node.val))
                queue.append(node.left)
                queue.append(node.right)
            else:
                ret.append('None')
        return '[' + ','.join(ret) + ']'

    def deserialize(self, data: str):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        if data == '[]':
            return None
        vals, index = data[1:-1].split(','), 1
        root = TreeNode(vals[0])
        queue = collections.deque()
        queue.append(root)
        while queue:
            node: TreeNode = queue.popleft()
            if vals[index] != 'None':
                node.left = TreeNode(int(vals[index]))
                queue.append(node.left)
            index += 1
            if vals[index] != 'None':
                node.right = TreeNode(int(vals[index]))
                queue.append(node.right)
            index += 1
        return root
# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))
# @lc code=end
